Change `dim` to `axis` for cosine_distance

yongtang opened the pull request:
This fix changes  `dim` to `axis` for cosine_distance so that the args are consistent with other methods in TensorFlow.

The backward-compatibility has been maintained in the fix.

This fix fixes #8205.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Change `dim` to `axis` for cosine_distance

This fix changes  `dim` to `axis` for cosine_distance
so that the args are consistent with other methods in TensorFlow.

The backward-compatibility has been maintained in the fix.

This fix fixes 8205.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Change `dim` to `axis` for tf.losses.cosine_distance

so that args are consistent with other TensorFlow methods.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

yifeif commented:
@martinwicke for api change approval.

yongtang commented:
The PR has been rebased and pushed to address the merge conflict.

drpngx commented:
Jenkins, test this please.

@martinwicke good to go?

yongtang commented:
The Jenkins failure is related to `//tensorflow/tools/api/tests:api_compatibility_test`. Will update the PR once api changes are approved.

sb2nov commented:
@martinwicke ping?

martinwicke commented on tensorflow/contrib/losses/python/losses/loss_ops.py:
```diff
@@ -633,7 +635,7 @@ def cosine_distance(
+    axis: The dimension along which the cosine distance is computed.
     weights: Coefficients for the loss a scalar, a tensor of shape
       [batch_size] or a tensor whose shape matches `predictions`.
     scope: The scope for the operations performed in computing the loss.
```
Please include dim in the docstring (the linter will complain if you don't). Can you make sure to look at the docstring of the function that the decorator returns to make sure that dim is appropriately marked as deprecated (and if not, put "(deprecated)" in its description)?

yongtang replied:
Thanks @martinwicke. The docstring has been updated with `dim` at the last line (match order of the API). Also dim is marked as deprecated correctly:
```
ubuntu@ubuntu:~$ python
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> print tf.losses.cosine_distance.__doc__
Adds a cosine-distance loss to the training procedure. (deprecated arguments)

SOME ARGUMENTS ARE DEPRECATED. They will be removed in a future version.
Instructions for updating:
dim is deprecated, use axis instead

Note that the function assumes that `predictions` and `labels` are already
unit-normalized.

Args:
  labels: `Tensor` whose shape matches 'predictions'
  predictions: An arbitrary matrix.
  axis: The dimension along which the cosine distance is computed.
  weights: Optional `Tensor` whose rank is either 0, or the same rank as
    `labels`, and must be broadcastable to `labels` (i.e., all dimensions must
    be either `1`, or the same as the corresponding `losses` dimension).
  scope: The scope for the operations performed in computing the loss.
  loss_collection: collection to which this loss will be added.
  reduction: Type of reduction to apply to loss.
  dim: The old (deprecated) name for `axis`.

Returns:
  Weighted loss float `Tensor`. If `reduction` is `NONE`, this has the same
  shape as `labels`; otherwise, it is scalar.

Raises:
  ValueError: If `predictions` shape doesn't match `labels` shape, or
    `axis`, `labels`, `predictions` or `weights` is `None`.
>>> exit()
ubuntu@ubuntu:~$ 
```

martinwicke commented on tensorflow/contrib/losses/python/losses/loss_ops.py:
```diff
@@ -645,8 +647,12 @@ def cosine_distance(
+      raise ValueError("Cannot specify both 'axis' and 'dim'")
+    axis = dim
+  if axis is None and dim is None:
+    raise ValueError("`axis` and `dim` cannot both be None.")
```
Can you change the error message here to "You must specify 'axis'."

Since dim is deprecated, no need to mention it.

yongtang replied:
Thanks. Done.

martinwicke commented on tensorflow/python/ops/losses/losses_impl.py:
```diff
@@ -242,7 +245,7 @@ def cosine_distance(
+    axis: The dimension along which the cosine distance is computed.
     weights: Optional `Tensor` whose rank is either 0, or the same rank as
       `labels`, and must be broadcastable to `labels` (i.e., all dimensions must
       be either `1`, or the same as the corresponding `losses` dimension).
```
Same here, add dim to docstring

yongtang replied:
Thanks. Done.

martinwicke commented on tensorflow/python/ops/losses/losses_impl.py:
```diff
@@ -256,10 +259,14 @@ def cosine_distance(
+      raise ValueError("Cannot specify both 'axis' and 'dim'")
+    axis = dim
+  if axis is None and dim is None:
+    raise ValueError("`axis` and `dim` cannot both be None.")
```
and don't mention dim here.

yongtang replied:
Done.

martinwicke commented:
Otherwise looks good, please run the compatibility golden update.

yongtang committed:
Update API goldens and address review feedback

This commit updates API goldens so that
`//tensorflow/tools/api/tests:api_compatibility_test`
could pass. Review feedback has also been addressed.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @martinwicke for the review and sorry for the late reply (was out of town until this week). The PR has been updated with API goldens rerun. Please take a look.

martinwicke commented:
Jenkins, test this please.

martinwicke commented:
Unrelated failure. @rmlarsen is the lingalg test known to be flaky?

drpngx commented:
Jenkins, test this please.

For reference,
```
//tensorflow/python/kernel_tests:linalg_ops_test                        TIMEOUT in 105.5s
```
